### PR TITLE
Timeline improvements when navigating with permalinks.

### PR DESCRIPTION
**Pull request by ganfra**
## Content

This PR tries to improve how the timeline behave when there is a click to permalink : 
- waits for indexer to be processed at least once, so we avoid calling focusTimeline if the items is already cached
- tries to center the focused event as much as possible
- avoid focusLive being called too early and weird jumps

This is still not perfect, but works better than previously.

## Motivation and context
Better UX when navigating with permalink.

## Tests

- Open a room
- Click on permalink (try with a non loading event, so you can have the forward pagination).

## Tested devices

- [ ] Physical
- [ ] Emulator
- OS version(s):

## Checklist

- [ ] Changes have been tested on an Android device or Android emulator with API 23
- [ ] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/element-hq/element-x-android/blob/develop/CONTRIBUTING.md#accessibility
- [ ] Pull request is based on the develop branch
- [ ] Pull request title will be used in the release note, it clearly define what will change for the user
- [ ] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [ ] You've made a self review of your PR

### CODE REVIEW

File: features/messages/impl/src/test/kotlin/io/element/android/features/messages/impl/timeline/TimelinePresenterTest.kt [hunk 1]
```diff
@@ -670,7 +670,7 @@ import kotlin.time.Duration.Companion.seconds
         timelineItemIndexer: TimelineItemIndexer = TimelineItemIndexer(),
     ): TimelinePresenter {
         return TimelinePresenter(
-            timelineItemsFactoryCreator = aTimelineItemsFactoryCreator(),
+            timelineItemsFactoryCreator = aTimelineItemsFactoryCreator(timelineItemIndexer),
```
Comment (bmarty) [inline features/messages/impl/src/test/kotlin/io/element/android/features/messages/impl/timeline/TimelinePresenterTest.kt:673]: ❤️

File: libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/utils/LazyListState.kt [hunk 2]
```diff
@@ -35,3 +37,33 @@ fun LazyListState.isScrollingUp(): Boolean {
...
+    }
+
+    // await for the first layout.
+    scroll { }
+    layoutInfo.resolveItemOffsetToCenter(index)?.let {
+        // Item is already visible, just scroll to center.
+        animateScrollToItem(index, it)
+        return
+    }
+    // Item is not visible, jump to to it centered.
```
Comment (bmarty) [inline libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/utils/LazyListState.kt:62]:
```suggestion
    // Item is not visible, jump to it centered.
```

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/TimelinePresenter.kt [hunk 3]
```diff
@@ -135,6 +136,12 @@ class TimelinePresenter @AssistedInject constructor(
                         )
                     } else {
                         newEventState.value = NewEventState.None
+                        if (event.firstIndex == 0) {
+                            val forwardPaginationStatus = timelineController.currentPaginationStatus(Timeline.PaginationDirection.FORWARDS)
+                            if (!forwardPaginationStatus.hasMoreToLoad) {
+                                timelineController.focusOnLive()
+                            }
```
Comment (bmarty) [inline features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/timeline/TimelinePresenter.kt:143]: It's weird to get data from the controller, to then ask the controller to do the action.

Just suggesting: maybe the added method `fun currentPaginationStatus` can be removed and this `if` block could be replaced by:
```kotlin
timelineController.maybeFocusOnLive(event.firstIndex)
```

**Review by bmarty — APPROVED** [review]
Thanks. The fix for the pinned event navigation when it's the last event is working as expected. No more loader in this case.

**Comment by frebib** [comment]
I tested to see if this fixes #2775 and it's definitely different, but it seems to still behave unintuitively

**Pull request closed without merging**